### PR TITLE
OpenMP multi-threading of skin and infill

### PR DESCRIPTION
**Pull request by Ghostkeeper**
Implements CURA-3330.

### CODE REVIEW

File: src/FffPolygonGenerator.cpp [hunk 1]
```diff
@@ -31,6 +32,17 @@ namespace cura
...
+    {
+#pragma omp master
+        {
+#ifdef _OPENMP
+            log("OpenMP enabled, number of threads used: %u\n", omp_get_num_threads());
+#else
+            log("OpenMP disabled\n");
+#endif
+        }
+    }
```
Comment (Ghostkeeper) [inline src/FffPolygonGenerator.cpp]: Why is this logging in the `generateAreas` function? Wouldn't you want this in a (very) central location, such as Main or FffProcessor?

File: src/FffPolygonGenerator.cpp [hunk 2]
```diff
@@ -345,12 +357,24 @@ void FffPolygonGenerator::processBasicWallsSkinInfill(SliceDataStorage& storage,
     
     
     // walls
-    for (unsigned int layer_number = 0; layer_number < mesh.layers.size(); layer_number++)
+    int processed_layer_count = 0;
+#pragma omp parallel for default(none) shared(mesh_layer_count, mesh, inset_skin_progress_estimate, processed_layer_count) schedule(dynamic)
+    for(unsigned int layer_number = 0; layer_number < mesh.layers.size(); layer_number++)
```
Comment (Ghostkeeper) [inline src/FffPolygonGenerator.cpp]: Code style: Space after for keyword. Fixed that for you in 2d9a3c83efe6b6d59c6314ba12243fd6260382be.

File: src/FffPolygonGenerator.cpp [hunk 3]
```diff
@@ -345,12 +357,24 @@ void FffPolygonGenerator::processBasicWallsSkinInfill(SliceDataStorage& storage,
     
     
     // walls
+    int processed_layer_count = 0;
```
Comment (Ghostkeeper) [inline src/FffPolygonGenerator.cpp]: `layer_number` is an `unsigned int`, while this is `int`.

(Not blocking.)

File: src/FffPolygonGenerator.cpp [hunk 4]
```diff
@@ -486,6 +527,14 @@ void FffPolygonGenerator::processDerivedWallsSkinInfill(SliceMeshStorage& mesh)
     }
 }
 
+/*
+ * This function is executed in a parallel region based on layer_nr.
+ * When modifying make sure any changes does not introduce data races.
+ *
+ * generateInsets only reads and writes data for the current layer
```
Comment (Ghostkeeper) [inline src/FffPolygonGenerator.cpp]: This documentation reports on some `generateInsets` function. Probably a copy-paste mistake. I'm removing it in 6861020c6b1282a9db28f4c3eeea460ff9651245.

File: src/FffPolygonGenerator.cpp (hunk 1 above)
Comment (BagelOrb) [inline src/FffPolygonGenerator.cpp]: No. You should call `omp_get_num_threads` within an o.p parallel region, since the value might differ each time.

Comment (BagelOrb) [inline src/FffPolygonGenerator.cpp]: I could improve on the text being shown though.

Comment (BagelOrb) [inline src/FffPolygonGenerator.cpp]: fixed

**Pull request merged**